tools/ops/policystream.py

```diff
@@ -0,0 +1,610 @@
+            self.policy.provider_name,
+            self.policy.resource_type,
+            self.date.isoformat(),
+            self.commit.author.name,
+            str(self.commit.id)[:6])
+
+    def data(self, indent=2):
+        d = {
+            'change': CHANGE_TYPE[self.kind].lower(),
+            'repo_uri': self.repo_uri,
+            'policy': {
+                'data': dict(self.policy.data),
+                'file': self.policy.file_path,
+            },
+            'commit': {
+                'id': str(self.commit.id),
+                'message': self.commit.message,
+                'author': str(self.commit.author.name),
+                'email': str(self.commit.author.email),
+                'date': self.date.isoformat(),
+            }
+        }
+        if self.previous:
+            d['previous'] = {
+                'data': dict(self.previous.data),
+                'file': self.previous.file_path}
+        return d
+
+
+class CollectionDelta(object):
+    """Iterator over changes between two policy collections.
+
+    With a given by url associated to a give commit.
+    """
+
+    change = PolicyChange
+
+    def __init__(self, prev, curr, commit=None, repo_uri=None):
+        self.prev = prev
+        self.curr = curr
+        self.commit = commit
+        self.repo_uri = repo_uri
+
+    def delta(self):
+        removed = set(self.prev.keys()) - set(self.curr.keys())
+        added = set(self.curr.keys()) - set(self.prev.keys())
+
+        for r in removed:
+            yield PolicyChange(self.prev[r], self.repo_uri, self.commit, ChangeType.REMOVE)
+
+        for a in added:
+            yield PolicyChange(self.curr[a], self.repo_uri, self.commit, ChangeType.ADD)
+
+        for extant in set(self.curr.keys()).intersection(set(self.prev.keys())):
+            if self.curr[extant].data != self.prev[extant].data:
+                yield PolicyChange(
+                    self.curr[extant], self.repo_uri, self.commit,
+                    ChangeType.MODIFIED, self.prev[extant])
+            elif self.curr[extant].file_path != self.prev[extant].file_path:
+                yield PolicyChange(
+                    self.curr[extant], self.repo_uri, self.commit,
+                    ChangeType.MOVED, self.prev[extant])
+
+
+class Policy(BasePolicy):
+    """Policy that tracks its file origin."""
+    def __init__(self, data, options, file_path):
+        self.file_path = file_path
+        self.data = data
+        self.options = options
+
+
+class PolicyCollection(BaseCollection):
+    """Policy collection supporting collection modifications."""
+
+    policy_class = Policy
+
+    def __init__(self, policies=None, options=None):
+        self.policies = policies or []
+        self.options = options or Config.empty()
+        self.pmap = {p.name: p for p in self.policies}
+
+    def __getitem__(self, key):
+        return self.pmap[key]
+
+    def add(self, p):
+        assert p.name not in self.pmap
```

Comment:
the assert here (and in remove likely) causes errors when running against repos that don't squash and merge commits as the commits show up as consecutive "adds" of policies

Reply:
the asserts are valid, we should have correctly detected the logical policy change at this layer, i've added the detection of this re-add as modify up to the git commit layer when looking at changed files.